README
JHOVE - JSTOR/Harvard Object Validation Environment
Copyright 2003-2012 by JSTOR and the President and Fellows of Harvard College
JHOVE is made available under the GNU Lesser General Public License (LGPL;
see the file LICENSE for details)

Rev. 1.11, 2013-09-29

JHOVE (the JSTOR/Harvard Object Validation Environment, pronounced "jhove")
is an extensible software framework for performing format identification,
validation, and characterization of digital objects.

o Format identification is the process of determining the format to which a
  digital object conforms: "I have a digital object; what format is it?"
o Format validation is the process of determining the level of compliance of a
  digital object to the specification for its purported format: "I have an
  object purportedly of format F; is it?"
o Format characterization is the process of determing the format-specific
  significant properties of an object of a given format: "I have an object of
  format F; what are its salient properties?"

These actions are frequently necessary during routine operation of digital
repositories and for digital preservation activities.

The output from JHOVE is controlled by output handlers. JHOVE uses an
extensible plug-in architecture; it can be configured at the time of its
invocation to include whatever specific format modules and output handlers
that are desired. The initial release of JHOVE includes modules for
arbitrary byte streams, ASCII and UTF-8 encoded text, AIFF and WAVE audio,
GIF, JPEG, JPEG 2000, TIFF, and PDF; and text and XML output handlers.

The JHOVE project is a collaboration of JSTOR and the Harvard University
Library.  Development of JHOVE was funded in part by the Andrew W. Mellon
Foundation.  JHOVE is made available under the GNU Lesser General Public
License (LGPL; see the file LICENSE for details).

JHOVE is currently being maintained by indpendent developers.

REQUIREMENTS

1. Java J2SE 1.5
(JHOVE was originally implemented using the Sun J2SE SDK 1.4.1 and has
been tested to work with 1.5)

2. If you would like to compile the JHOVE source code, then
Apache Ant, a Java-based build tool <http://ant.apache.org/> is necessary.
Note that the JAVA_HOME environment variable must be set appropriately for
Ant to work properly.
(JHOVE was implemented and tested using Ant 1.5.1.)

DISTRIBUTION

The JHOVE distribution package includes:

    jhove/                                 # JHOVE home directory
          COPYING                          # GNU Lesser General Public License
          LICENSE                          # JHOVE license information
          README
          RELEASENOTES                     # JHOVE release notes
          bin/
                    jhove.jar              # JHOVE API package
                    jhove-handler.jar      # Standard output handler package
                    jhove-module.jar       # Standard module package
                    JhoveApp.jar           # JHOVE command line application
                    JhoveView.jar          # JHOVE with Swing GUI front-end
          build.xml                        # Ant configuration file
          classes/
                    build.xml              # Ant configuration file
                    edu/ ...               # JHOVE API packages
                    ADump.*                # AIFF dump utility class
                    GDump.*                # GIF dump utility class
                    Jhove.*                # JHOVE main class
                    JDump.*                # JPEG dump utility class
                    J2Dump.*               # JPEG 2000 dump utility class
                    PDump.*                # PDF dump utility class
                    TDump.*                # TIFF dump utility class
                    UserHome.*             # user.home property utility class
                    WDump.*                # WAVE dump utility class
          conf/
                   jhove.conf              # JHOVE configuration file
                   jhove.xsd               # JHOVE output schema
                   jhoveConfig.xsd         # JHOVE configuration file schema
          doc/
                   *.html                  # API documentation
                   ...
          examples/                        # Sample files
                   ascii/ ...
                   gif/   ...
                   jpeg/  ...
                   jpeg2000/ ...
                   pdf/   ...
                   tiff/  ...
                   utf-8/ ...
          adump*                           # AIFF dump Bourne shell driver
          adump.bat*                       # AIFF dump DOS shell driver script
          gdump*                           # GIF dump Bourne shell driver
          gdump.bat*                       # GIF dump DOS shell driver script
          jdump*                           # JPEG dump Bourne shell driver
          jdump.bat*                       # JPEG dump DOS shell driver script
          j2dump*                          # JPEG 2000 dump Bourne shell driver
          j2dump.bat*                      # JPEG 2000 dump DOS shell driver
          jhove.tmpl*                      # Template for JHOVE Bourne shell driver script
          jhove_bat.tmpl*                  # Template for JHOVE DOS shell driver script
          pdump*                           # PDF dump Bourne shell driver
          pdump.bat*                       # PDF dump DOS shell driver script
          tdump*                           # TIFF dump Bourne shell driver
          tdump.bat*                       # TIFF dump DOS shell driver script
          userhome*                        # user.home Bourne shell driver
          userhome.bat*                    # user.home DOS shell driver script
          wdump*                           # WAVE dump Bourne shell driver
          wdump.bat*                       # WAVE dump DOS shell driver script

INSTALLATION

Edit the configuration file, jhove/conf/jhove.conf, and set the absolute
pathname of the JHOVE home directory and the temporary directory (in which
temporary files are created):

    <jhoveHome>jhove-home-directory</jhoveHome>
    <tempDirectory>temporary-directory</tempDirectory>

The JHOVE home directory is the top-most directory in the distribution TAR
or ZIP file.  On Unix systems, "/var/tmp" is an appropriate temporary
directory; on Windows, "C:\Temp".  For example, if the distribution TAR
file is disaggregated on a Unix system in the directory "/users/stephen/
projects", then the configuration file should read:

    <jhoveHome>/users/stephen/projects/jhove</jhoveHome>
    <tempDirectory>/var/tmp</jhoveHome>

In the JHOVE home directory, copy the JHOVE Bourne shell driver script
template, "jhove.tmpl", to "jhove" (or the equivalent Windows shell 
script, "jhove_bat.tmpl" to "jhove.bat"), and set the
JHOVE home directory, Java home directory, and Java interpreter:

    JHOVE_HOME=jhove-home-directory
    JAVA_HOME=java-home-directory
    JAVA=java-interpreter

The JAVA_HOME property should provide the absolute pathname of the Java
runtime or SDK installation; JAVA should provide the absolute pathname of the
Java interpreter.  For example:

    JHOVE_HOME=/users/stephen/projects/jhove
    JAVA_HOME=/usr/local/j2re1.4.1_02
    JAVA=$JAVA_HOME/bin/java

In the DOS shell driver script, jhove.bat, the equivalent three
variables are:

    SET JHOVE_HOME=jhove-home-directory
    SET JAVA_HOME=java-home-directory
    SET JAVA=%JAVA_HOME%\bin\java

For example:

    SET JHOVE_HOME="C:\Program Files\jhove"
    SET JAVA_HOME="C:\Program Files\java\j2re1.4.1_02"
    SET JAVA=%JAVA_HOME%\bin\java

The quotation marks are necessary because of the embedded space characters.
On Windows platforms it may also be necessary to add the Java bin subdirectory
to the System PATH environment variable:

    PATH=C:\Program Files\java\j2re1.4.1_02\bin;...

(For information on setting a Windows environment variable, consult your local
documentation or system administrator.)

USAGE

    java Jhove [-c config] [-m module] [-h handler] [-e encoding] [-H handler]
               [-o output] [-x saxclass] [-t tempdir] [-b bufsize]
               [-l loglevel] [[-krs] dir-file-or-uri [...]]

where -c config   Configuration file pathname
      -m module   Module name
      -h handler  Output handler name (defaults to TEXT)
      -e encoding Character encoding used by output handler (defaults to UTF-8)
      -H handler  About handler name
      -o output   Output file pathname (defaults to standard output)
      -x saxclass SAX parser class (defaults to J2SE default)
      -t tempdir  Temporary directory in which to create temporary files
      -b bufsize  Buffer size for buffered I/O (defaults to J2SE 1.4 default)
      -l loglevel Logging level
      -k          Calculate CRC32, MD5, and SHA-1 checksums
      -r          Display raw data flags, not textual equivalents
      -s          Format identification based on internal signatures only
      dir-file-or-uri Directory or file pathname or URI of formated content
                      stream

All named modules and output handlers must be found on the Java CLASSPATH at
the time of invocation.  The JHOVE driver script, jhove/jhove, automatically
sets the CLASSPATH and invokes the Jhove main class:

    jhove [-c config] [-m module] [-h handler] [-e encoding] [-H handler]
          [-o output] [-x saxclass] [-t tempdir] [-b bufsize] [-l loglevel]
          [[-krs] dir-file-or-uri [...]]

The following additional programs are available, primarily for testing
and debugging purposes.  They display a minimally processed, human-readable
version of the contents of AIFF, GIF, JPEG, JPEG 2000, PDF, TIFF, and WAVE
files:

    java ADump  aiff-file
    java GDump  gif-file
    java JDump  jpeg-file
    java J2Dump jpeg2000-file
    java PDump  pdf-file
    java TDump  tiff-file
    java WDump  wave-file

For convenience, the following driver scripts are also available:

    adump  aiff-file
    gdump  gif-file
    jdump  jpeg-file
    j2dump jpeg2000-file
    pdump  pdf-file
    tdump  tiff-file
    wdump  wave-file

The JHOVE Swing-based GUI interface can be invoked from a command shell from
the jhove/bin sub-directory:

    java -jar JhoveView.jar -c <configFile>

where <configFile> is the pathname of the JHOVE configuration file.


RELEASENOTES
RELEASENOTES
JHOVE - JSTOR/Harvard Object Validation Environment
Copyright 2003-2009 by JSTOR and the President and Fellows of Harvard College
JHOVE is made available under the GNU Lesser General Public License (LGPL;
see the file LICENSE for details)

Versions 1.7 and beyond of JHOVE are no longer under the control of Harvard.

RELEASE NOTES FOR JHOVE 1.11

GENERAL

1. I've added lots of logging code. Calls at the FINE level and lower 
   don't show up no matter what I do, so I've put them at the INFO level.
   The level is set in JhoveBase.java.

2. All .bat and _bat.tmpl files now have CR-LF line endings. That is, they
   do in the gzip and zip archives you download. I'm not sure how
   SourceForge will treat files that you download individually,
   but hopefully it will have the sense to keep CR-LF when downloading
   to a Windows system. 

3. All .bat files now assume JHOVE_HOME is the directory from which they're
   run. They no longer try to set JAVA_HOME (which was still stuck in
   Java 1.4 and probably wasn't working for many people), instead assuming
   that the JAVA command is available on the command line.

4. All javac commands in build.xml files now specify source=1.5 for
   compatibility with more recent compilers.
  
5. gdumpwin.bat is deleted. It's redundant with gdump.bat and has bugs
   of its own.
   
PDF MODULE

1. Fix to PDF module, submitted by willp-bl, may reduce tendency 
   to run out of heap space on some files.

RELEASE NOTES FOR JHOVE 1.10

GENERAL

1. The amount of logging code has been increased, mostly at the
   DEBUG level.
   
2. Further work on generics in Java code.

3. JhoveView now checks for Java 1.5. Was previously allowing 1.4 even
   though it wouldn't work.

HTML MODULE

1. XHTML files are processed by the HTML module, which invokes the XML
   modules. In this case, the XML module doesn't have the parameters
   specified in the JHOVE configuration file and so won't use local
   copies of schemas. Starting with this version, the parameters of
   the HTML module are passed to the XML module when invoking it.
   However, this doesn't work properly (in either module) for a DTD
   that invokes additional DTDs by relative URLs. Such DTDs should
   be edited to use only absolute URLs.
   
PDF MODULE

1. Failure to get a page object number wasn't being handled cleanly,
   resulting in a report of an invalid document without an error message
   to explain it (SourceForge bug 49). This has been fixed.

2. The PDF module unnecessarily uses huge amounts of memory to build
   complex structure trees, when it doesn't need to keep the whole 
   tree in memory to validate it. In the new version, it uses memory
   more economically. This should result in the successful processing
   of some PDF files that ran out of memory or took hours to process before.
   
3. If an annotation isn't a dictionary object, report that explicitly.
   This happens with some otherwise good files; I can't find any warrant
   for it in the PDF spec.

4. Some efficiency improvements to PDF parser. Increased buffer size from 4K 
   to 64K. Made Parser.collapseObjectVector more efficient. Parser now
   returns pseudo-objects for array and dictionary end instead of throwing
   an exception.

5. Minor cleanup of error reporting.

6. If an object uses a compression scheme which JHOVE can't deal with, JHOVE
   will try to give a specific error message.

RELEASE NOTES FOR JHOVE 1.9

GENERAL

1. Jhove.java and JhoveView.java now get their version information from 
   JhoveBase.java. Before it was redundantly kept in three places, and 
   sometimes they didn't all get updated for a new release. Like in 1.8.
   
2. ConfigWriter was in the package edu.harvard.hul.ois.jhove.viewer, which 
   caused a NoClassDefFoundError if non-GUI configurations didn't include
   JhoveViewer.jar in the classpath. It's been moved to
   edu.harvard.hul.ois.jhove.

3. Added script packagejhove.sh and made md5.pl part of the CVS repository
   to make packaging for delivery easier.

4. jhove.bat now simply uses the Java command rather than requiring
   the user to set up the Java path.

5. JhoveView.jar and jhove (the top level shell script) are now forced
   by ant to be executable so there are no mistakes.

6. Warning message given on invalid buffer size string, and minimum 
   buffer size is 1024.

7. Configuration file code for adding handlers and giving init strings 
   to modules was an awful mess that never could have worked. 
   Major repairs done.
   
AIFF MODULE

1. If an AIFF file was found to be little-endian, the module instance
   would stay in little-endian mode for all subsequent files. This
   has been fixed.
   
TIFF MODULE

1. TIFF files that had strip or tile offsets but no corresponding byte
   counts were throwing an exception all the way to the top level. Now
   they're correctly being reported as invalid.
   
XML MODULE

1. Cleaned up reporting of schemas, Added some small classes to replace
   the use of string arrays for information structures. Made URI comparison
   for local schema parameter case-independent. Resolved conflict between
   "s" and "schema" parameters.

WAVE MODULE

1.  Some uncaught exceptions caused the module to throw all the way
    back to JhoveBase and not report any result for certain defective
    files. These now report the file as not well-formed. 


RELEASE NOTES FOR JHOVE 1.8

GENERAL

1. If JHOVE doesn't find a configuration file, it creates a default one.

2. Generics widely added to clean up the code.

3. build.xml files fixed to force compilation to Java 1.5.

4. Shell script "jhove" no longer makes you figure out where JAVA_HOME is.

PDF MODULE

1. Several errors in checking for PDF-A compliance were corrected. Aside from
   fixing some outright bugs, the Contents key for non-text Annotations is

classes/Jhove.java
/**********************************************************************
 * Jhove - JSTOR/Harvard Object Validation Environment
 * Copyright 2004-2007 by the President and Fellows of Harvard College
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or (at
 * your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
 * USA
 **********************************************************************/

import edu.harvard.hul.ois.jhove.*;
//import java.io.*;
import java.util.*;

public class Jhove
{
    /******************************************************************
     * PRIVATE CLASS FIELDS.
     ******************************************************************/

    /** Application name. */
    private static final String NAME = "Jhove";

    /** Application build date, YYYY, MM, DD. */
    private static final int [] DATE = JhoveBase.DATE;

    /** Application release number. */
    private static final String RELEASE = JhoveBase._release;

    /** Application invocation syntax. */
    private static final String USAGE = "java " + NAME + " [-c config] " +
        "[-m module] [-h handler] [-e encoding] [-H handler] [-o output] " +
	"[-x saxclass] [-t tempdir] [-b bufsize] [-l loglevel] [[-krs] " +
	"dir-file-or-uri [...]]";

    /** Copyright information. */
    private static final String RIGHTS =
	"Derived from software Copyright 2004-2011 " +
    "by the President and Fellows of Harvard College. " +
    "Version 1.7 and higher independently released. " +
	"Released under the GNU Lesser General Public License.";

    /******************************************************************
     * MAIN ENTRY POINT.
     ******************************************************************/

    public static void main (String [] args)
    {
        /* Make sure we have a satisfactory version of Java. */
        String version = System.getProperty ("java.vm.version");
        if (version.compareTo ("1.4.0") < 0) {
            //System.err.println (NAME + ": Java 1.4 or higher is required");
            System.out.println ("Java 1.4 or higher is required");
            System.exit (-1);
        }

        try {
    
    	    /**********************************************************
    	     * Initialize the application state object.
    	     **********************************************************/
    
    	    App app = new App (NAME, RELEASE, DATE, USAGE, RIGHTS);
    
    	    /**********************************************************
    	     * Retrieve the configuration file.
    	     **********************************************************/
    
    	    String configFile = JhoveBase.getConfigFileFromProperties ();
    	    String saxClass   = JhoveBase.getSaxClassFromProperties ();
    
    	    /* Pre-parse the command line for -c and -x config options. 
    	     * With Windows, we have to deal with quote marks on our own. With Unix,
    	     * the shell takes care of quotes for us. */
    	    boolean quoted = false;
    	    for (int i=0; i<args.length; i++) {
    		if (quoted) {
    		    int len = args[i].length ();
                if (args[i].charAt (len-1) == '"') {
                    quoted = false;
                }
    		}
    		else {
    		    if (args[i].equals ("-c")) {
                    if (i < args.length-1) {
                        configFile = args[++i];
                    }
                }
    		    else if (args[i].equals ("-x")) {
                    if (i <args.length-1) {
                        saxClass = args[++i];
                    }
                }
    		    else if (args[i].charAt (0) == '"') {
                    quoted = true;
                }
    		}
        }
    
    	    /**********************************************************
    	     * Initialize the JHOVE engine.
    	     **********************************************************/
    
    	    String encoding     = null;
    	    String tempDir      = null;
    	    int    bufferSize   = -1;
    
            String moduleName   = null;
            String handlerName  = null;
            String aboutHandler = null;
            String logLevel     = "SEVERE";
            String outputFile   = null;
            boolean checksum    = false;
            boolean showRaw     = false;
            boolean signature   = false;
            List<String> list   = new ArrayList<String> ();
    
    	 
                /**********************************************************
                 * Parse command line arguments:
                 *  -m module    Module name
                 *  -h handler   Output handler
    	     *  -e encoding  Output encoding
                 *  -H handler   About handler
                 *  -o output    Output file pathname
                 *  -t tempdir   Directory for temp files
                 *  -b bufsize   Buffer size for buffered I/O
                 *  -k           Calculate checksums
                 *  -r           Display raw numeric flags 
                 *  -s           Check internal signatures only
                 *  dirFileOrUri Directories, file pathnames, or URIs
    	     *
    	     * The following arguments were defined in previous
    	     * versions, but are now obsolete
                 *  -p param     OBSOLETE
                 *  -P param     OBSOLETE
    	     **********************************************************/
    
    	    quoted = false;
    	    StringBuffer filename = null;
    
            for (int i=0; i<args.length; i++) {
                if (quoted) {
                    int len = args[i].length ();
                    if (args[i].charAt (len-1) == '"') {
                        filename.append (" " + args[i].substring (0, len-1));
                        list.add (filename.toString ());
                        quoted = false;
                    }
                    else {
                        filename.append (" " + args[i]);
                    }
                }
                else {
        		    if (args[i].equals ("-c")) {
                        i++;
        		    }
        		    else if (args[i].equals ("-m")) {
                        if (i < args.length-1) {
                            moduleName = args[++i];
                        }
        		    }
        		    else if (args[i].equals ("-p")) {
                        // Obsolete -- but eat the next arg for compatibility
                        if (i <args.length-1) {
                            String moduleParam = args[++i];
                        }
                    }
        		    else if (args[i].equals ("-h")) {
                        if (i < args.length-1) {
                            handlerName = args[++i];
                        }
                    }
        		    else if (args[i].equals ("-P")) {
                        // Obsolete -- but eat the next arg for compatibility
                        if (i <args.length-1) {
                            String handlerParam = args[++i];
                        }
        		    }
        		    else if (args[i].equals ("-e")) {
                        if (i < args.length-1) {
                            encoding = args[++i];
                        }
        		    }
        		    else if (args[i].equals ("-H")) {
                        if (i < args.length-1) {
                            aboutHandler = args[++i];
                        }
                    }
        		    else if (args[i].equals ("-l")) {
                        if (i < args.length-1) {
                            logLevel = args[++i];
                        }
        		    }
        		    else if (args[i].equals ("-o")) {
                        if (i < args.length-1) {
                            outputFile = args[++i];
                        }
                    }
        		    else if (args[i].equals ("-x")) {
                        i++;
        		    }
        		    else if (args[i].equals ("-t")) {
                        if (i <args.length-1) {
                            tempDir = args[++i];
                        }
                    }
        		    else if (args[i].equals ("-b")) {
                        if (i <args.length-1) {
                            try {
                                bufferSize = Integer.parseInt (args[++i]);
                            }
                            catch (NumberFormatException e) {
                                System.err.println ("Invalid buffer size, using default.");
                            }
                        }
        		    }
        		    else if (args[i].equals ("-k")) {
                        checksum = true;
        		    }
        		    else if (args[i].equals ("-r")) {
                        showRaw = true;
        		    }
        		    else if (args[i].equals ("-s")) {
                        signature = true;
                    }
        		    else if (args[i].charAt (0) != '-') {
                        if (args[i].charAt (0) == '"') {
                            filename = new StringBuffer ();
                            filename.append (args[i].substring (1));
                            quoted = true;
                        }
                        else {
                            list.add (args[i]);
                        }
        		    }
        		}
            }
            if (quoted) {
                list.add (filename.toString ());
    	    }
    
            JhoveBase je = new JhoveBase ();
            je.setLogLevel (logLevel);
            je.init (configFile, saxClass);
            if (encoding == null) {
                encoding = je.getEncoding ();
            }
            if (tempDir == null) {
                tempDir = je.getTempDirectory ();
            }
            if (bufferSize < 0) {
                bufferSize = je.getBufferSize ();
            }
            Module module = je.getModule  (moduleName);
            if (module == null && moduleName != null) {
                System.out.println ("Module '" + moduleName + "' not found");
                System.exit (-1);
            }
            OutputHandler about   = je.getHandler (aboutHandler);
            if (about == null && aboutHandler != null) {
                System.out.println ("Handler '" + aboutHandler + "' not found");
                System.exit (-1);
            }
    	    OutputHandler handler = je.getHandler (handlerName);
            if (handler == null && handlerName != null) {
                System.out.println ("Handler '" + handlerName + "' not found");
                System.exit (-1);
            }
    	    String [] dirFileOrUri = null;
    	    int len = list.size ();
            if (len > 0) {
        		dirFileOrUri = new String [len];
        		for (int i=0; i<len; i++) {
        		    dirFileOrUri[i] = (String) list.get (i);
        		}
            }
    
    	    /**********************************************************
    	     * Invoke the JHOVE engine.
    	     **********************************************************/
                
            je.setEncoding (encoding);
            je.setTempDirectory (tempDir);
            je.setBufferSize (bufferSize);
            je.setChecksumFlag (checksum);
            je.setShowRawFlag (showRaw);
            je.setSignatureFlag (signature);
    	    je.dispatch (app, module, about, handler, outputFile, 
    			 dirFileOrUri);
    	}
        catch (Exception e) {
            e.printStackTrace (System.err);
            System.exit (-1);
        }
    }
    

}
